### DISCUSSION TITLE
Viewer Right Click Cut, Copy, Paste

### DISCUSSION THREAD

**Discussion by MsBlockCat** [opentoonz/opentoonz#6539] [Ideas]
### Issue Summary
The Xsheet has Cut, Copy, and Paste options when you right click, as most programs do. In the Viewer, however, although Cmd + X (Ctrl + X on Windows), Cmd + C (Ctrl + C on Windows), and Cmd + V (Ctrl + V on Windows) do work, the right click menu doesn't include Cut, Copy, and Paste

### Screenshot or Video Reference
Xsheet's right click menu (has Cut, Copy, and Paste options)
<img width="248" alt="Screen Shot 2020-03-28 at 12 03 04 AM" src="https://user-images.githubusercontent.com/37186692/77814448-8efc0b80-7087-11ea-989b-f592f2fcf33b.png">

Viewer's right click menu (no such options)
<img width="232" alt="Screen Shot 2020-03-28 at 12 04 48 AM" src="https://user-images.githubusercontent.com/37186692/77814469-c5d22180-7087-11ea-8175-41d58fe4f45a.png">

### Expected Results
I think that, as the Xsheet right click menu does and other programs usually do, Cut, Copy, and Paste should probably be at the top of the right click menu as they're frequently used

### Actual Results
There aren't any Cut, Copy, or Paste options in the Viewer's right click menu

### System Information
* **OpenToonz Version**: Recent nightly, 1.3
* **Operating System**: macOS Catalina (10.15.4), Windows 10
* **CPU**: Quad Core i5, Quad Core Ryzen 5 2500U
* **Memory**: 8 GB LPDDR3, 8 GB DDR4
* **Graphics Card**: Intel Iris Pro 645, Radeon Vega 8 Mobile

**Comment by RodneyBaker** [comment]
@MrBlockCat In the xsheet's Right Click menu the copying relates to cell exposures.
Can you clarify exactly what would be copied/cut when using this option via a Viewer?
I must assume it would be something selected by the user.

**Comment by MsBlockCat** [comment]
I think the best way to go about this would be to just replicate the functionality the keyboard shortcuts (Ctrl/Cmd + X/C/V) already do when a Viewer’s selected, right?

**Comment by RodneyBaker** [comment]
Not sure.  That's why I asked.  :)
There are a few other Right Click additions I'd love to see but I'm sure these follow each users personal workflow preferences closely.  For example, currently I have been using the Vector Optimization options a lot and often find myself wanting to just Right Click to trigger those actions. **

A full fix that allowed for everyone's needs would likely be to make the Right Click menu be customizable.  As the UI's menus in that area aren't part of LESS/QSS I'd think that could be a major undertaking.

**Someone recently posted about one of the utilties they used to add additional menus and that might be a good solution for those users with a increased need for customization

**Comment by MsBlockCat** [comment]
Fair point! I personally think the Cut/Copy/Paste right click options should replicate that functionality, then, as I think that’s what people would expect to happen

I agree it would be ideal if the right click menus could be customized, but I have no idea how big of a project that would be so if it is that’s unfortunate! The utility sounds interesting, I presume it isn’t some kind of open source one we could have built in with credit?

**Comment by manongjohn** [comment]
Something to keep in mind: what commands are available in the viewer depends on the tool you are using at the time. Some commands may not make sense as right-click options depending on the tool being used. From a tool's perspective, those commands make sense for the Selection tool. Not sure I particularly care to see that in right-click menu option when they aren't valid for the tool.

**Comment by RodneyBaker** [comment]
> The utility sounds interesting, I presume it isn’t some kind of open source one we could have built in with credit?

I did a cursory look for the post that referred to the utility but didn't find it.
We should take a survey of similar utilities/hardware being used, what is available... and determine what is standardized/cross platform to maximum extent possible.   

Regarding this feature request itself...
Now you've got me curious about where the Right Click menu options are in the source.

Aside:  I really like Right Click options but I'm my own worst enemy in this regard because for most commands I could just be using a shortcut key.  I think this relates to me preferring to either be fully using mouse/pen or keyboard at any given time and not both simultaneously.  The most successful examples I've seen are of people that have mastered using both keyboard and pen/mouse simultaneously.

Added:  Very good point @manongjohn !

**Comment by MsBlockCat** [comment]
I agree that the options in the right click menu should be completely removed when the commands can't be used! The right click menu is a little long, removing options when they don't make sense is good in my opinion

**Comment by MsBlockCat** [comment]
It appears the Viewer's right click options are at toonz/sources/toonz/sceneviewercontextmenu.cpp ([a link to it](https://github.com/opentoonz/opentoonz/blob/233fcf81e35bf032991e3f85b291a3d0ce1bbedd/toonz/sources/toonz/sceneviewercontextmenu.cpp))

**Comment by MsBlockCat** [comment]
@RodneyBaker I've added a $10 bounty to add this: https://www.bountysource.com/issues/90840065-viewer-right-click-cut-copy-paste

**Comment by MsBlockCat** [comment]
@RodneyBaker Sorry to bother- but you seem to have forgotten this one! :)

**Comment by RodneyBaker** [comment]
Recommend anyone with a current bounty placed on a feature or bugfix review the current status and try to resolve the bounty at earliest convenience.

**Comment by MsBlockCat** [comment]
@RodneyBaker Agreed, I'll most likely try and refund the remaining bounties I have open soon due to an extended period of inactivity with them. Would you prefer I kept the related issues open, or should they be closed when I revoke the associated bounties? Thanks!

**Comment by RodneyBaker** [comment]
I don't particularly have a preference.
If the issue is still an active one then it should generally stay open and if there is no longer a need it can be closed.
Another way to deal with some issues is to attach it to a similar issue (assuming one can be found).

BTW - It's good to hear from you!  :)

**Comment by MsBlockCat** [comment]
Thanks! I've removed the remaining bounties of mine; It seems I can't remove the Bounty tag myself on my issues, I'd appreciate it if you could on them for accuracy's sake when you have the time

**Comment by RodneyBaker** [comment]
Done.  Thanks @MsBlockCat
